stripped/colabs/Implementacja_NearestNeighbours_stripped.py

```diff
@@ -0,0 +1,177 @@
+spark.conf.set("park.driver.maxResultSize", "80g")
+
+spark.conf.set('spark.sql.execution.arrow.enabled', 'true')
+
+
+drive.mount('/content/drive')
+
+
+# Wczytanie zbioru sampled w celu pobrania nazw kolumn
+sampled = pd.read_csv('/content/drive/MyDrive/BigMess/NASA/sampled_NASA_200k.csv')
+
+# Utworzenie schematu określającego typ zmiennych
+schemat = StructType()
+for i in sampled.columns:
+  if i == "Date":
+    schemat = schemat.add(i, IntegerType(), True)
+  else:
+    schemat = schemat.add(i, FloatType(), True)
+
+
+%%time
+# Wczytanie zbioru Nasa w sparku
+nasa = spark.read.format('csv').option("header", True).schema(schemat).load('/content/drive/MyDrive/BigMess/NASA/NASA.csv')
+nasa.show(5)
+
+nasa.createOrReplaceTempView("nasa")
+
+nasa = (
+    nasa
+    .withColumn('Year', (F.col('Date') / 100).cast('int'))
+    .withColumn('Month', F.col('Date') % 100)
+    .drop('Date')
+)
+nasa.show(5)
+
+
+nasa_coords = spark.sql("""SELECT DISTINCT lat, lon FROM nasa""")
+nasa_coords.collect()
+
+
+"""
+### **2. Implementacja algorytmu kNN:**
+"""
+
+# function searches for points that lie within a (euclidean) ball of size *radius* around the query points
+# if optional argument *k* is given then function searches for (at most) k nearest points that lie within
+# a (euclidean) ball of size *radius* around the query points
+
+#point = (latitude, longitude)
+def kRadiusNN(df: SparkDataFrame, radius: float, point: Tuple[float, float], k: Optional[int]=None) -> pd.DataFrame :
+  assert (25.0625<= point[0] <=52.9375) and (-124.9375<= point[1] <=-67.0625 ), 'Wrong coordinates (out of range)'
+
+  df = df.collect()
+  neighbours_pd = pd.DataFrame({"lon":[], "lat":[], 'dist':[]})
+
+  for row in df:
+
+     lon = row['lon']
+     lat = row['lat']
+     dist = geodesic((lat, lon), point).km
+     if dist <= radius:
+        new_row = {'lon': lon , 'lat': lat, 'dist' : dist}
+        neighbours_pd.loc[len(neighbours_pd)] = new_row
+
+  if k and (k < len(neighbours_pd)):
+
+      neighbours_pd = neighbours_pd.sort_values('dist', ascending=True)
+      if not (neighbours_pd.at[k-1,'dist']==neighbours_pd.at[k,'dist']):  #checking if there is no tie (more neighbours with the same distance)
+         neighbours_pd = neighbours_pd.iloc[:k]
+      else:
+         raise Exception("Unable to determine k nearest neighbours: more neighbours with the same distance")
+
+  return(neighbours_pd)
+
+
+kRadiusNN(nasa_coords, 20, (35, -110), 5)
+
+
+# weighted: if True then function will weight points by the inverse of their distance (in this case, closer neighbours of
+# a query point will have a greater influence than neighbors which are further away).
+
+def predict_class(df: SparkDataFrame, point: Tuple[float, float], radius: float, label_column_name: str,
+                  k: Optional[int]=None, weighted: Optional[bool]=False) -> bool :
+ if k:
+    neighbours = kRadiusNN(df, radius, point, k)
+ else:
+    neighbours = kRadiusNN(df, radius, point)
```

Comment:
Co to robi?

Reply:
Jak wyżej: wyjaśniam w korespondencji na MS Teams.